stl/inc/ranges

```diff
@@ -6837,6 +6837,602 @@ namespace ranges {
+
+    private:
+        tuple<_ViewTypes...> _M_Views;
+    };
+
+    template <typename... _RangeTypes>
+    zip_view(_RangeTypes&&...) -> zip_view<views::all_t<_RangeTypes>...>;
+
+    template <input_range... _RangeTypes>
+    inline constexpr bool enable_borrowed_range<zip_view<_RangeTypes...>> = (enable_borrowed_range<_RangeTypes> && ...);
+
+    namespace _Zip {
+        template <typename _ResultType, typename _LHSTuple, typename _RHSTuple>
+        struct _Misc_iterator_sentinel_noexcept_solver {};
+
+        template <typename _ResultType, typename... _LHSTupleTypes, typename... _RHSTupleTypes>
+        struct _Misc_iterator_sentinel_noexcept_solver<_ResultType, tuple<_LHSTupleTypes...>,
+            tuple<_RHSTupleTypes...>> {
+            static constexpr bool _Is_get_smallest_distance_noexcept =
+                (noexcept(_ResultType{_STD declval<_LHSTupleTypes>() - _STD declval<_RHSTupleTypes>()}) && ...)
+                && (noexcept(
+                        _ResultType{_RANGES distance(_STD declval<_LHSTupleTypes>(), _STD declval<_RHSTupleTypes>())})
+                    && ...)&& noexcept(_STD declval<_ResultType>() < _STD declval<_ResultType>())
+                && is_nothrow_copy_assignable_v<_ResultType>;
+        };
+
+        template <typename... _LHSTupleTypes, typename... _RHSTupleTypes>
+        struct _Misc_iterator_sentinel_noexcept_solver<bool, tuple<_LHSTupleTypes...>, tuple<_RHSTupleTypes...>> {
+            static constexpr bool _Is_evaluate_equality_noexcept =
+                (noexcept(_STD declval<_LHSTupleTypes>() == _STD declval<_RHSTupleTypes>()) && ...);
+        };
+
+        template <typename _ResultType, typename _LHSItrOrSentTuple, typename _RHSItrOrSentTuple>
+            requires(tuple_size_v<decay_t<_LHSItrOrSentTuple>> == tuple_size_v<decay_t<_RHSItrOrSentTuple>>)
+        _NODISCARD constexpr _ResultType _Get_smallest_iterator_sentinel_distance(_LHSItrOrSentTuple&& _Lhs_tuple,
+            _RHSItrOrSentTuple&& _Rhs_tuple) noexcept(_Misc_iterator_sentinel_noexcept_solver<_ResultType,
+            decay_t<_LHSItrOrSentTuple>, decay_t<_RHSItrOrSentTuple>>::_Is_get_smallest_distance_noexcept);
+
+        template <typename _LHSItrOrSentTuple, typename _RHSItrOrSentTuple>
+            requires(tuple_size_v<decay_t<_LHSItrOrSentTuple>> == tuple_size_v<decay_t<_RHSItrOrSentTuple>>)
+        _NODISCARD constexpr bool _Evaluate_iterator_sentinel_equality(_LHSItrOrSentTuple&& _Lhs_tuple,
+            _RHSItrOrSentTuple&& _Rhs_tuple) noexcept(_Misc_iterator_sentinel_noexcept_solver<bool,
+            decay_t<_LHSItrOrSentTuple>, decay_t<_RHSItrOrSentTuple>>::_Is_evaluate_equality_noexcept);
+    } // namespace _Zip
+
+    template <input_range... _ViewTypes>
+        requires _Zip::_Allow_views<_ViewTypes...>
+    template <bool _IsConst>
+    class zip_view<_ViewTypes...>::_Iterator : public _Iterator_category_container<_Iterator<_IsConst>> {
+    private:
+        friend class zip_view<_ViewTypes...>;

```

Comment:
We conventionally use C++11 "extended `friend` syntax" whenever possible, i.e. omitting `class` and saying just `friend zip_view<_ViewTypes...>;`. This has the advantage of being unable to declare anything new, which is one of the most confusing things about `friend`. Occurs below. (Unfortunately this can't be used when granting templated friendship, or we'd use it even more. Here, `zip_view<_ViewTypes...>` is a specific single type in this specialization, so the syntax works.)